pkg/etcdcli/cached_members_test.go

```diff
@@ -0,0 +1,314 @@
+		{
+			"one unhealthy non voting member and two healthy voting members",
+			asLearner(defaultEtcdMembers(), 3),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 2, Unhealthy: 1}),
+			fakeEtcdMembers(1, 2),
+		},
+		{
+			"all unhealthy voting members",
+			defaultEtcdMembers(),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 0, Unhealthy: 3}),
+			emptyEtcdMembers(),
+		},
+		{
+			"all unhealthy non voting members",
+			asLearner(defaultEtcdMembers(), 1, 2, 3),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 0, Unhealthy: 3}),
+			emptyEtcdMembers(),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client, err := NewFakeEtcdClient(tc.etcdMembersList, tc.clientOpts)
+			require.NoError(t, err)
+			memberCache := newMemberCache(client, 1*time.Millisecond)
+
+			currentMembers, err := memberCache.HealthyVotingMembers(context.Background())
+			require.Equal(t, tc.expEtcdMembers, currentMembers)
+		})
+	}
+}
+
+func TestCachedMembers_UnhealthyMembers(t *testing.T) {
+	testCases := []struct {
+		name            string
+		etcdMembersList []*etcdserverpb.Member
+		clientOpts      FakeClientOption
+		expEtcdMembers  []*etcdserverpb.Member
+	}{
+		{
+			"all healthy members",
+			defaultEtcdMembers(),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 3, Unhealthy: 0}),
+			emptyEtcdMembers(),
+		},
+		{
+			"one unhealthy member and two healthy members",
+			defaultEtcdMembers(),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 2, Unhealthy: 1}),
+			fakeEtcdMembers(3),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client, err := NewFakeEtcdClient(tc.etcdMembersList, tc.clientOpts)
+			require.NoError(t, err)
+			memberCache := newMemberCache(client, 1*time.Millisecond)
+
+			currentMembers, err := memberCache.UnhealthyMembers(context.Background())
+			require.Equal(t, tc.expEtcdMembers, currentMembers)
+		})
+	}
+}
+
+func TestCachedMembers_UnhealthyVotingMembers(t *testing.T) {
+	testCases := []struct {
+		name            string
+		etcdMembersList []*etcdserverpb.Member
+		clientOpts      FakeClientOption
+		expEtcdMembers  []*etcdserverpb.Member
+	}{
+		{
+			"all healthy voting members",
+			defaultEtcdMembers(),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 3, Unhealthy: 0}),
+			emptyEtcdMembers(),
+		},
+		{
+			"one unhealthy voting member and two healthy voting members",
+			defaultEtcdMembers(),
+			WithFakeClusterHealth(&FakeMemberHealth{Healthy: 2, Unhealthy: 1}),
+			fakeEtcdMembers(3),
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			client, err := NewFakeEtcdClient(tc.etcdMembersList, tc.clientOpts)
+			require.NoError(t, err)
+			memberCache := newMemberCache(client, 1*time.Millisecond)
+
+			currentMembers, err := memberCache.UnhealthyMembers(context.Background())
+			require.Equal(t, tc.expEtcdMembers, currentMembers)
+		})
+	}
+}
+
+func TestCachedMembers_MemberHealth(t *testing.T) {
```

Comment:
@tjungblu  this test is failing, seems the memberHealth is not responding to the fake

Reply:
it is responding, it just always returns all members with a status: healthy true/false.

Reply:
done 👍🏽